data/transactions/logic/assembler.go

```diff
@@ -1545,9 +1548,16 @@ func (ops *OpStream) trackStack(args StackTypes, returns StackTypes, instruction
 	}
 }
 
-// splitTokens breaks tokens into two slices at the first semicolon.
-func splitTokens(tokens []string) (current, rest []string) {
-	for i, token := range tokens {
+// splitTokens breaks tokens into two slices at the first semicolon and expands macros along the way.
+func splitTokens(ops *OpStream, tokens []string) (current, rest []string) {
+	for i := 0; i < len(tokens); i++ {
+		token := tokens[i]
+		replacement, ok := ops.macros[token]
+		if ok {
+			tokens = append(tokens[0:i], append(replacement, tokens[i+1:]...)...)
+			i--
+			continue
+		}
```

Comment:
I think you have to loop or expand recursively, expanding your expansions. I don't think the current expansion will handle:
```
#define A X
#define X 1
int A; assert
```

which means, by the way, that you can work your cycle checking into this expansion. In fact, that's the point - the reason you want to detect cycles is exactly because expansion should continue until it stops, which it won't do if you have a cycle.

Reply:
I'm fairly confident it works since we decrement the index, but I've also just now tested the example you listed and it worked. The reason I put the cycle check where the macros are defined was because it's somewhat easier to generate a good error there and we can only error once instead of constantly erroring for the same macro cycle

Reply:
OK, I missed the purpose of the i--. That seems good. Could you // backup to process the expansion or similar?

This does mean that expansion to nothing isn't allowed. I know you check for that in #define, so I think it's probably fine.

But then again, it could be useful. Imagine, for example
```
#define DEBUG#1 byte #1; log
```
And the author changes that to:

```
#define DEBUG#1
```
before assembling for deployment

Reply:
I don't think I care about expansion to an empty.  The writer could do:
```
#define DEBUG#1 ;
```
instead.